### PR TITLE
Preliminary reading from views support (Issue #21)

### PR DESCRIPTION
**Pull request by davidrabinowitz**

### CODE REVIEW

File: src/main/scala/com/google/cloud/spark/bigquery/BigQueryUtil.scala [hunk 1]
```diff
@@ -65,4 +66,7 @@ object BigQueryUtil {
   }
 
   def noneIfEmpty(s: String): Option[String] =  Option(s).filterNot(_.trim.isEmpty)
+
+  def convertAndThrow(error: BigQueryError) : Unit =
+    throw new BigQueryException(UNKNOWN_CODE, error.getMessage, error)
```
Comment (kmjung) [inline src/main/scala/com/google/cloud/spark/bigquery/BigQueryUtil.scala]: Is it appropriate to always use UNKNOWN_CODE here?

File: src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala [hunk 2]
```diff
@@ -127,23 +150,90 @@ private[bigquery] class DirectBigQueryRelation(
...
+      log.debug(s"querySql is $querySql")
+      destinationTableCache.get(querySql, DestinationTableBuilder(querySql))
+    } else {
+      // use the default one
+      table
+    }
+  }
+
+  def createTableFromQuery(querySql: String): TableInfo = {
+    val destinationTable = createDestinationTable
```
Comment (kmjung) [inline src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala]: How are these temporary tables deleted after creation?

File: src/main/scala/com/google/cloud/spark/bigquery/BigQueryUtil.scala (hunk 1 above)
Comment (davidrabinowitz) [reply src/main/scala/com/google/cloud/spark/bigquery/BigQueryUtil.scala]: The code is modeled after [BigQueryException.translateAndThrow()](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryException.java#L98). As [BigQueryError](https://github.com/googleapis/google-cloud-java/blob/master/google-cloud-clients/google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/BigQueryError.java) does not contain the code, I see no other option.

File: src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala (hunk 2 above)
Comment (davidrabinowitz) [reply src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala]: I thought that it was set to 24 hours automatically... The connector now updates the destination table's expiration time immediately after it is created

File: src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala [hunk 3]
```diff
@@ -127,23 +150,97 @@ private[bigquery] class DirectBigQueryRelation(
...
+  // return empty if no filters are used
+  def createWhereClause(filters: Array[Filter]): Option[String] = {
+    val filtersString = DirectBigQueryRelation.compileFilters(filters)
+    BigQueryUtil.noneIfEmpty(filtersString)
+  }
+
+  def createDestinationTable: TableId = {
+    val uuid = UUID.randomUUID()
+    val name = s"_sbc_${uuid.getMostSignificantBits.toHexString}${uuid.getLeastSignificantBits.toHexString}"
+    TableId.of(tableId.getProject, tableId.getDataset, name)
```
Comment (kmjung) [inline src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala]: You might consider making this configurable -- in some cases, it won't be appropriate to put the temporary table in the same dataset as the source table. Beam, for example, creates not only a temporary table, but a temporary dataset as well which it deletes at the end of pipeline execution.

Comment (davidrabinowitz) [reply src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala]: Added the configuration options, see https://github.com/GoogleCloudPlatform/spark-bigquery-connector/pull/83/commits/1cb67b5f60c38fd9575050b53ffd7d784dfc3b50

File: src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala (hunk 2 above)
Comment (davidrabinowitz) [reply src/main/scala/com/google/cloud/spark/bigquery/direct/DirectBigQueryRelation.scala]: See https://github.com/GoogleCloudPlatform/spark-bigquery-connector/pull/83/commits/bdabb4216ea95818bf43569ad7d2e53e61d24c84

**Comment by kmjung** [comment]
Thank you, David!

**Pull request merged**